### PR TITLE
Feature 6025 travel entry forms

### LINKED ISSUE #6025: DEA TravelEntry form [3]
### Situation Description

Concept issue for travel entries: #5489

### Feature Description

All elements are only visible on German systems

#### Edit form
* Add a DEA view class that consists of two tabs ("Travel Entry" and "Person") and that is opened when accessing a specific travel entry from the directory
* Read-only for users without the `TRAVEL_ENTRY_EDIT` user right
* Edit rights are based on the responsible jurisdiction and point of entry jurisdiction of the travel entry
* Add a DEAForm that contains UI components for the TravelEntryDto fields
  * Reporting user is read only
  * `Recovered`, `Vaccinated` and `Tested negative` are checkboxes and hidden if no disease has been selected
  * The value of `deaContent` is parsed into a list of text fields; the `CampaignFormBuilder` class can be used as a reference on how to do this, but probably can't be re-used; create a `DEAFormBuilder` class instead that reads the JSON and can also build JSON based on the fields again when the user saves
* Edit form needs to include archive and delete options (only for users with the `TRAVEL_ENTRY_ARCHIVE` and `TRAVEL_ENTRY_DELETE` user right, respectively)
* Responsible jurisdiction and point of entry fields should behave like the responsible jurisdiction and place of stay fields for cases, i.e. add a checkbox "Point of entry jurisdiction differs from responsible jurisdiction" that shows the point of entry region and district fields, point of entry always needs to be required and either depends on the point of entry jurisdiction fields if available or falls back to the responsible jurisdiction fields otherwise

#### Create form
* Reuse the edit form for the create form
* Add person fields, same as when creating a case
* Add a "New travel entry" button to the travel entry directory that opens this form
* Only accessible for users with the `TRAVEL_ENTRY_CREATE` user right
* Use duplicate person detection on save; duplicate travel entry detection is not needed
* Hide the "New travel entry" button if no travel entry has been imported into the system yet; if at least one travel entry is available, build the text fields for the DEAFormBuilder based on the `deaContent` captions of the latest travel entry

### PR DESCRIPTION
**Pull request by sergiupacurariu**
Fixes #6025

### CODE REVIEW

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
* Deleted travel entries still appear in the directory - could you fix that in this PR please?
* Same for archived entries; there also needs to be a dropdown to toggle between archived, active and all entries (as for other directories)
* Some of the fields only have a very small width (see screenshot below)
* Checking "point of entry jurisdiction differs from responsible jurisdiction" copies the responsible district over to the point of entry district, but not the region (either both or none are fine); moreover, when deselecting it again, the point of entry is removed as well and can no longer be selected because it apparently still relies on the empty point of entry region and district fields instead of the responsible jurisdiction fields (both in the create and edit forms)
* The "Land" and "Risikoklasse" fields are not saved, i.e. not written to the database; might have something to do with the captions being not unique, but I don't think it's a database problem because it works when importing data

image

File: sormas-api/src/main/java/de/symeda/sormas/api/Disease.java [hunk 1]
```diff
@@ -181,4 +184,6 @@ public int keyCompareTo(StatisticsGroupingKey o) {
 
 		return this.toString().compareTo(o.toString());
 	}
+
+	public static List<Disease> IN_DISEASE_LIST = Arrays.asList(Disease.values());
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/java/de/symeda/sormas/api/Disease.java:188]: changed to "DISEASE_LIST"

File: sormas-backend/src/main/resources/sql/sormas_schema.sql [hunk 2]
```diff
@@ -7672,4 +7672,10 @@ ALTER TABLE events_history ADD COLUMN eventidentificationsource varchar(255);
 
 INSERT INTO schema_version (version_number, comment) VALUES (385, 'Event identification source (#5526)');
 
+-- 2021-07-19 DEA Travel entry form #6025
+ALTER TABLE travelentry ALTER COLUMN deacontent TYPE json USING deacontent::json;
+ALTER TABLE travelentry_history ALTER COLUMN deacontent TYPE json USING deacontent::json;
+
+INSERT INTO schema_version (version_number, comment) VALUES (386, 'DEA Travel entry form #6025');
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-backend/src/main/resources/sql/sormas_schema.sql:7679]: updated the message including the recommended message

File: sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntriesView.java [hunk 3]
```diff
@@ -49,6 +54,13 @@ public TravelEntriesView() {
 				Window popupWindow = VaadinUiUtil.showPopupWindow(new TravelEntryImportLayout());
 				popupWindow.setCaption(I18nProperties.getString(Strings.headingImportTravelEntries));
 			}));
+
+			List<String> travelEntries = FacadeProvider.getTravelEntryFacade().getAllUuids();
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntriesView.java:58]: changed to count method

File: sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryDataForm.java [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+	}
+
+	@Override
+	protected String createHtmlLayout() {
+		return HTML_LAYOUT;
+	}
+
+	@Override
+	protected void addFields() {
+//		if (person == null || disease == null) {
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryDataForm.java:129]: removed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryDataForm.java [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+//			return;
+//		}
+
+		Label travelEntryDataHeadingLabel = new Label(I18nProperties.getString(Strings.headingTravelEntryData));
+		travelEntryDataHeadingLabel.addStyleName(H3);
+		getContent().addComponent(travelEntryDataHeadingLabel, TRAVEL_ENTRY_HEADING_LOC);
+
+		addField(de.symeda.sormas.api.travelentry.TravelEntryDto.REPORT_DATE, DateField.class);
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryDataForm.java:137]: update all of them in the class

File: sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryDataForm.java [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+		UserProvider currentUserProvider = UserProvider.getCurrent();
+		JurisdictionLevel userJurisditionLevel =
+			currentUserProvider != null ? UserRole.getJurisdictionLevel(currentUserProvider.getUserRoles()) : JurisdictionLevel.NONE;
+		if (userJurisditionLevel == JurisdictionLevel.COMMUNITY) {
+			regionCombo.setReadOnly(true);
+			districtCombo.setReadOnly(true);
+		} else if (userJurisditionLevel == JurisdictionLevel.HEALTH_FACILITY) {
+			regionCombo.setReadOnly(true);
+			districtCombo.setReadOnly(true);
+		}
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryDataForm.java:203]: rewrite the logic

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
- [x] Change captions of the "All", "Active" and "Archived" options to "All travel entries", "Archived travel entries", etc. (instead of "... cases")
- [ ] The dropdown is always cleared when entering the view, even though the previous (or default) selection applies; the selection should be displayed in the dropdown
- [ ] Reload the travel entry list after importing
- [x] Interestingly, I still have one case where the text field only takes part of the available width

**Review by MateStrysewske — APPROVED** [review]

**Pull request merged**